cgstats.js

```diff
@@ -126,37 +126,74 @@ app.get('/search*', function(req, res) {
       });
     });
   } else if (optimizations.indexOf(game) != -1) {
+    getStats('LeaderboardsRemoteService/getFilteredPuzzleLeaderboard',
+      [game, , , {"active" : true, "column" : "keyword", "filter" : player}])
+      .then(function (response) {
+        var requests = response.users
+          .filter(function(rank) {
+            return (rank.pseudo || '').toLowerCase() == player.toLowerCase();
+          })
+          .map(function(rank) {
+            return getStats('LeaderboardsRemoteService/getFilteredPuzzleLeaderboard',
+              [game, , , {"active" : true, "column" : "SLANGUAGE", "filter" : rank.programmingLanguage}]);
+          });
+        Promise.all(requests)
+          .then(function(responses) {
+            var users = responses.map(function(response) {
+              return response.users;
+            })
+            .reduce(function(a, b) {
+              return a.concat(b);
+            }, []);
+            return {users: users, programmingLanguages: response.programmingLanguages};
+          })
+          .then(function(response) {
+            res.type('json').set({
+              'Access-Control-Allow-Origin' : 'http://cgstats.magusgeek.com'
+            }).send(JSON.stringify({
+              player : player,
+              stats : compileOptimizationStats(response, player),
+              mode : 'optim'
+            })).end();
+          })
+          .catch(function(error) {
+            res.status(500).end();
+          });
+      })
+      .catch(function(error) {
+        res.status(500).end();
+      });
+  }
+});
+
+app.listen(9888);
+
+// *****************************
+
+function getStats(resource, parameters) {
+  return new Promise(function(resolve, reject) {
```

Comment:
I didn't want to introduce another dependency so I simply wrapped existing request into new promise.
Promise required for async downloading of stats for all languages that the user solved the puzzle with.

Reply:
cgstats run in a nodejs 8+. Don't worry for Promise :p